module/VuFind/src/VuFind/Auth/AbstractShibboleth.php

```diff
@@ -0,0 +1,372 @@
+ * @author   Bernd Oberknapp <bo@ub.uni-freiburg.de>
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @author   Vaclav Rosecky <vaclav.rosecky@mzk.cz>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+abstract class AbstractShibboleth extends AbstractBase
+{
+    /**
+     * Header name for entityID of the IdP that authenticated the user.
+     */
+    const DEFAULT_IDPSERVERPARAM = 'Shib-Identity-Provider';
+
+    /**
+     * Header name of the internal session key assigned to the
+     * session associated with the request.
+     */
+    const SHIB_SESSION_ID = "Shib-Session-ID";
+
+    /**
+     * This is array of attributes which $this->authenticate()
+     * method should check for.
+     *
+     * WARNING: can contain only such attributes, which are writeable to user table!
+     *
+     * @var array attribsToCheck
+     */
+    protected $attribsToCheck = [
+        'cat_username', 'cat_password', 'email', 'lastname', 'firstname',
+        'college', 'major', 'home_library'
+    ];
+
+    /**
+     * Session manager
+     *
+     * @var \Laminas\Session\ManagerInterface
+     */
+    protected $sessionManager;
+
+    /**
+     * Constructor
+     *
+     * @param \Laminas\Session\ManagerInterface $sessionManager Session manager
+     */
+    public function __construct(\Laminas\Session\ManagerInterface $sessionManager)
+    {
+        $this->sessionManager = $sessionManager;
+    }
+
+    /**
+     * Get the URL to establish a session (needed when the internal VuFind login
+     * form is inadequate).  Returns false when no session initiator is needed.
+     *
+     * @param string $target Full URL where external authentication method should
+     * send user after login (some drivers may override this).
+     *
+     * @return bool|string
+     */
+    public function getSessionInitiator($target)
+    {
+        $config = $this->getConfig();
+        $shibTarget = $config->Shibboleth->target ?? $target;
+        $append = (strpos($shibTarget, '?') !== false) ? '&' : '?';
+        // Adding the auth_method parameter makes it possible to handle logins when
+        // using an auth method that proxies others.
+        $sessionInitiator = $config->Shibboleth->login
+            . '?target=' . urlencode($shibTarget)
+            . urlencode($append . 'auth_method=Shibboleth');
+
+        if (isset($config->Shibboleth->provider_id)) {
+            $sessionInitiator = $sessionInitiator . '&entityID=' .
+                urlencode($config->Shibboleth->provider_id);
+        }
+
+        return $sessionInitiator;
+    }
+
+    /**
+     * Has the user's login expired?
+     *
+     * @return bool
+     */
+    public function isExpired()
+    {
+        $proxy = $this->getConfig()->Shibboleth->proxy ?? false;
+        return ($proxy) ? isset($_SERVER[$this->normalize(self::SHIB_SESSION_ID)])
```

Comment:
Should we pass the request object into isExpired() via checkForExpiredCredentials() so we can avoid directly accessing superglobals here? Would that enable us to use the standard getAttribute() support method here instead?

Also, I notice that the current code before refactoring does an `empty` check rather than an `isset` check. Is it better to do that more forgiving check, or is the change done here for a reason?